api/src/main/java/org/pmiops/workbench/cohortbuilder/SearchGroupItemQueryBuilder.java

```diff
@@ -0,0 +1,578 @@
+            ? VALUE_SOURCE_CONCEPT_ID
+            : VALUE_AS_CONCEPT_ID),
+        conceptIdParam,
+        OperatorUtils.getSqlOperator(attribute.getOperator()),
+        operandsParam);
+  }
+
+  /** Helper method to build the operand sql expression. */
+  private static String getOperandsExpression(
+      Map<String, QueryParameterValue> queryParams, Attribute attribute) {
+    String operandsParam1 =
+        addQueryParameterValue(
+            queryParams, QueryParameterValue.float64(new Double(attribute.getOperands().get(0))));
+    String valueExpression;
+    if (attribute.getOperator().equals(Operator.BETWEEN)) {
+      String operandsParam2 =
+          addQueryParameterValue(
+              queryParams, QueryParameterValue.float64(new Double(attribute.getOperands().get(1))));
+      valueExpression = operandsParam1 + AND + operandsParam2;
+    } else {
+      valueExpression = operandsParam1;
+    }
+    return valueExpression;
+  }
+
+  /** Collect all child nodes per specified search parameters. */
+  private static Set<Long> childConceptIds(
+      Map<SearchParameter, Set<Long>> criteriaLookup, List<SearchParameter> params) {
+    Set<Long> out = Sets.newHashSet();
+    for (SearchParameter param : params) {
+      if (param.getGroup()) {
+        out.addAll(criteriaLookup.get(param));
+      }
+      if (param.getConceptId() != null) {
+        // not all SearchParameter have a concept id, so attributes/modifiers
+        // are used to find matches in those scenarios.
+        out.add(param.getConceptId());
+      }
+    }
+    return out;
+  }
+
+  /** Helper method to build modifier sql if needed. */
+  private static String buildModifierSql(
+      String baseSql, Map<String, QueryParameterValue> queryParams, List<Modifier> modifiers) {
+    String ageDateAndEncounterSql = getAgeDateAndEncounterSql(queryParams, modifiers);
+    // Number of Occurrences has to be last because of the group by
+    String occurrenceSql =
+        buildOccurrencesSql(queryParams, getModifier(modifiers, ModifierType.NUM_OF_OCCURRENCES));
+    return String.format(MODIFIER_SQL_TEMPLATE, baseSql + ageDateAndEncounterSql) + occurrenceSql;
+  }
+
+  /**
+   * Helper method to build all modifiers together except occurrences since it has to be last
+   * because of the group by.
+   */
+  private static String getAgeDateAndEncounterSql(
+      Map<String, QueryParameterValue> queryParams, List<Modifier> modifiers) {
+    List<Modifier> ageDateAndEncounterModifiers = new ArrayList<>();
+    ageDateAndEncounterModifiers.add(getModifier(modifiers, ModifierType.AGE_AT_EVENT));
+    ageDateAndEncounterModifiers.add(getModifier(modifiers, ModifierType.EVENT_DATE));
+    ageDateAndEncounterModifiers.add(getModifier(modifiers, ModifierType.ENCOUNTERS));
+    StringBuilder modifierSql = new StringBuilder();
+    for (Modifier modifier : ageDateAndEncounterModifiers) {
+      if (modifier != null) {
```

Comment:
nit: to reduce nesting, I would invert this and continue

Reply:
Done.